org-roam-db.el

```diff
@@ -544,33 +571,63 @@ database, see `org-roam-db-sync' command."
     (cond
      (enabled
       (add-hook 'find-file-hook  #'org-roam-db-autosync--setup-file-h)
+      (org-roam-db-autosync--update-method)
       (add-hook 'kill-emacs-hook #'org-roam-db--close-all)
-      (advice-add #'rename-file :after  #'org-roam-db-autosync--rename-file-a)
-      (advice-add #'delete-file :before #'org-roam-db-autosync--delete-file-a)
       (org-roam-db-sync))
      (t
       (remove-hook 'find-file-hook  #'org-roam-db-autosync--setup-file-h)
+      (org-roam-db-autosync--update-method)
       (remove-hook 'kill-emacs-hook #'org-roam-db--close-all)
-      (advice-remove #'rename-file #'org-roam-db-autosync--rename-file-a)
-      (advice-remove #'delete-file #'org-roam-db-autosync--delete-file-a)
       (org-roam-db--close-all)
       ;; Disable local hooks for all org-roam buffers
       (dolist (buf (org-roam-buffer-list))
         (with-current-buffer buf
           (remove-hook 'after-save-hook #'org-roam-db-autosync--try-update-on-save-h t)))))))
 
-;;;###autoload
-(defun org-roam-db-autosync-enable ()
-  "Activate `org-roam-db-autosync-mode'."
-  (org-roam-db-autosync-mode +1))
-
-(defun org-roam-db-autosync-disable ()
-  "Deactivate `org-roam-db-autosync-mode'."
-  (org-roam-db-autosync-mode -1))
-
-(defun org-roam-db-autosync-toggle ()
-  "Toggle `org-roam-db-autosync-mode' enabled/disabled."
-  (org-roam-db-autosync-mode 'toggle))
+(defvar org-roam-db-autosync--filenotify-descriptors (list)
+  "An alist mapping watched Org-roam directories to `filenotify' descriptor.")
+
+(defun org-roam-db-autosync--update-method ()
+  "Setup `org-roam-db-autosync-update-method' dependently on the mode's state."
+  (cond
+   (org-roam-db-autosync-mode ; enabled
+    (pcase org-roam-db-autosync-update-method
+      ('filenotify
+       (cl-pushnew
+        (cons org-roam-directory
+              (file-notify-add-watch org-roam-directory '(change)
+                                     #'org-roam-db-autosync--filenotify-update))
+        org-roam-db-autosync--filenotify-descriptors))
```

Comment:
`file-notify-add-watch` isn't recursive, so we need to think through how to handle subdirectories. The current logic should handle multiple `org-roam-directory` just fine... I think (untested).

Reply:
> file-notify-add-watch isn't recursive, so we need to think through how to handle subdirectories.

ahh... we can watch every directory that's currently in `org-roam-directory`, and then on event, if a directory is created we can use the callback to create more file watchers. Not sure if it's efficient/worth it anymore though.

Reply:
I mean, we kind of halfway there, so might as well try to handle this? Though if you think it's not worth the hassle to continue, then we can stop. 

If you wish to continue, then we can update [the current facility](https://github.com/org-roam/org-roam/blob/b8a66deae99e0feab3106d7b7ec36af1631f2dff/org-roam.el#L233-L307) to only list subdirectories given a special parameter, and then add something like `org-roam-list-(sub?)directories`. Though it shouldn't include patterns like `^\.`, so we wouldn't add file watchers to e.g. `git` related stuff. And to handle different `org-roam-directory` values we can use `add-variable-watcher` or `org-roam-find-file-hook`.

Reply:
I think i'll write a helper library for this and publish it to MELPA.

Reply:
https://github.com/jethrokuan/filenotify-recursive here it is... i think it works for now.

Reply:
I'll take a look and might even try to integrate it there, before it will be added to Melpa. Do you think listing directories with Elisp won't be a bottleneck? I could add support for external finders (similarly to what we have in Org-roam) to the package, if you think it would worth it.

Reply:
Huh, surprisingly `fnr--subdirectories-recursively` isn't slow. Don't think it will need external finders, at least until someone will show an actual demand for that, so I'll just integrate it to this PR as given.

Reply:
> I'll try to integrate it before it will be added to Melpa. But before doing so, do you think listing directories with Elisp won't be a problem? I could add support for external finders (similarly to what we have in Org-roam) to the package, if you think it would worth it.

I went ahead and did it already, and am noticing just one issue: It's not very smart in excluding hidden folders (like .git), so it does a lot of unnecessary watching.

Reply:
Can you expose a `defvar` for the [predicate](https://github.com/jethrokuan/filenotify-recursive/blob/049de0b15d8478c2ffed008a580e7bc36c6b4a59/filenotify-recursive.el#L65)? Then we can use some dumb heuristic to determine whether each directory is hidden or not.